File: ProgrammingErlang/creditcard.erl
-module(creditcard).
-export([type/1, luhn/1]).
-import(lists, [map/2, zip/2, reverse/1]).
-include_lib("eunit/include/eunit.hrl").
-compile(export_all).

onetwos(N) ->
    onetwos(N, []).

onetwos(0, Acc) ->
    Acc;
onetwos(N, Acc) ->
    onetwos(N - 1, [2 - (N rem 2)|Acc]).

onetwos_test_() ->
    [?_assertMatch([], onetwos(0)),
     ?_assertMatch([1], onetwos(1)),
     ?_assertMatch([1,2,1], onetwos(3)),
     ?_assertMatch([1,2,1,2], onetwos(4))].

luhn_digit(N) when N > 9 ->
    N - 9;
luhn_digit(N) ->
    N.

luhn_checksum(Digits) ->
    lists:sum(map(fun({X, Y}) -> luhn_digit((X - $0) * Y) end,
		  zip(Digits, reverse(onetwos(length(Digits)))))).

luhn(Digits) ->
    luhn_checksum(Digits) rem 10 =:= 0.

luhn_test_() ->
    [?_assertMatch(3, luhn_checksum("3")),
     ?_assertMatch(7, luhn_checksum("23")),
     ?_assertMatch(7, luhn_checksum("80")),
     ?_assert(luhn("4408041234567893")),
     ?_assert(luhn("4438241234567893")),
     ?_assert(not luhn("4417123456789112"))].

luhn() ->
    fun(Digits) ->
	    luhn(Digits)
    end.

length_is(Length) when is_integer(Length) ->
    fun(Digits) ->
	    length(Digits) =:= Length
    end;
length_is(Lengths) when is_list(Lengths) ->
    satisfy_any(map(fun length_is/1, Lengths)).

length_is_test_() ->
    P = length_is(3),
    [?_assert(P("123")),
     ?_assert(not P("12"))].

length_is_one_of_test_() ->
    P = length_is([3, 5]),
    [?_assert(P("123")),
     ?_assert(P("12345")),
     ?_assert(not P("12"))].

begin_with([H|_]=Prefix) when is_integer(H) ->
    fun(Digits) ->
	    lists:prefix(Prefix, Digits)
    end;
begin_with([H|_]=Prefixes) when is_list(H) ->
    satisfy_any(map(fun begin_with/1, Prefixes)).

begin_with_range(S, E) ->
    Range = map(fun integer_to_list/1,
		lists:seq(list_to_integer(S), list_to_integer(E))),
    begin_with(Range).

begin_with_test_() ->
    P = begin_with("34"),
    [?_assert(P("345")),
     ?_assert(not P("356"))].

begin_with_one_of_test_() ->
    P = begin_with(["34", "44"]),
    [?_assert(P("345")),
     ?_assert(P("445")),
     ?_assert(not P("356"))].

satisfy_all(Predicates) ->
    fun(Digits) ->
	    lists:foldl(fun(P, Acc) -> P(Digits) and Acc end,
			true, Predicates)
    end.

satisfy_any(Predicates) ->
    fun(Digits) ->
	    lists:foldl(fun(P, Acc) -> P(Digits) or Acc end,
			false, Predicates)
    end.
    
card_type(Name, Discriminant) ->
    {Name, Discriminant}.

card_types() ->
    [card_type(amex,
	       satisfy_all([begin_with(["34","37"]), length_is(15)])),
     card_type(discover,
	       satisfy_all([begin_with("6011"), length_is(16)])),
     card_type(mastercard,
	       satisfy_all([begin_with_range("51", "55"), length_is(16)])),
     card_type(visa, satisfy_all([begin_with("4"), length_is([13,16])]))].

guess_type(Digits, CardTypes) ->
    [Name || {Name, Discriminant} <- CardTypes, Discriminant(Digits)].

type(Digits) ->
    type(Digits, card_types()).
type(Digits, CardTypes) ->
    case guess_type(Digits, CardTypes) of
	[] ->
	    unknown;
	[Name] ->
	    Name;
	Names ->
	    {error, Names}
    end.

type_test_() -> 
    [?_assertMatch(visa, type("4408041234567893")),
     ?_assertMatch(visa, type("4417123456789112")),
     ?_assertMatch({error, [acme, visa]}, type("4408041234567893", [card_type(acme, satisfy_all([]))|card_types()]))].

File: ProgrammingErlang/CreditCard.hs
import Char
import Test.HUnit

type Digit = Char

oneTwos :: [Int]
oneTwos = oneTwos 1
    where
      oneTwos 1 = 1:oneTwos 2
      oneTwos 2 = 2:oneTwos 1

luhnChecksum :: [Digit] -> Int
luhnChecksum digits = sum (map weight pairs)
    where
      pairs = zip (reverse (map digitToInt digits)) oneTwos
      weight (x, y) = if x * y > 9 then x * y - 9 else x * y

luhn :: [Digit] -> Bool
luhn digits = rem (luhnChecksum digits) 10 == 0

lengthIs :: Int -> [a] -> Bool
lengthIs expected xs = length xs == expected

beginWith :: Eq a => [a] -> [a] -> Bool
beginWith expected xs = take (length expected) xs == expected

satisfyAny :: [a -> Bool] -> a -> Bool
satisfyAny predicates x = or (map (\p -> p x) predicates)

satisfyAll :: [a -> Bool] -> a -> Bool
satisfyAll predicates x = and (map (\p -> p x) predicates)

lengthIsOneOf :: [Int] -> [a] -> Bool
lengthIsOneOf es xs = satisfyAny (map lengthIs es) xs

beginWithOneOf :: Eq a => [[a]] -> [a] -> Bool
beginWithOneOf es xs = satisfyAny (map beginWith es) xs

beginWithRange :: [Digit] -> [Digit] -> [Digit]-> Bool
beginWithRange s e xs = beginWithOneOf (range s e) xs
    where
      range :: [Digit] -> [Digit] -> [[Digit]]
      range s e = map show [(read s :: Int)..(read e :: Int)]

guessType :: [Digit] -> [String]
guessType digits = [name | (name, discriminant) <- types, discriminant digits]
    where
      types = [("AMEX", satisfyAll [beginWithOneOf ["34", "37"], lengthIs 15]),
               ("Discover", satisfyAll [beginWith "6011", lengthIs 16]),
               ("MasterCard", satisfyAll [beginWithRange "51" "55", lengthIs 16]),
               ("Visa", satisfyAll [beginWith "4", lengthIsOneOf [13,16]])]

tests = test [
         "1" ~: take 1 oneTwos ~?= [1],
         "12" ~: take 2 oneTwos ~?= [1,2],
         "121" ~: take 3 oneTwos ~?= [1,2,1],
         "Luhn checksum" ~: luhnChecksum "1" ~?= 1,
         "Luhn checksum" ~: luhnChecksum "12" ~?= 4,
         "Luhn checksum" ~: luhnChecksum "80" ~?= 7,
         "Luhn checksum is valid" ~: luhn "34" ~?= True,
         "Luhn checksum is invalid" ~: luhn "123" ~?= False,
         "Length is" ~: lengthIs 3 "124" ~?= True,
         "Length is not" ~: lengthIs 4 "124" ~?= False,
         "Begins with" ~: beginWith "432" "4321" ~?= True,
         "Doesn't begin with" ~: beginWith "123" "124" ~?= False,
         "Satisfies any" ~: satisfyAny [t_p, f_p] "1" ~?= True,
         "Doesn't satisfy any" ~: satisfyAny [f_p, f_p] "1" ~?= False,
         "Satisfies all" ~: satisfyAll [t_p, t_p] "1" ~?= True,
         "Doesn't satisfy all" ~: satisfyAll [t_p, f_p] "1" ~?= False,
         "It's Visa" ~: guessType "4417123456789112" ~?= ["Visa"]
        ]
    where
      t_p x = True
      f_p x = False
